File: NoTomorrow/Features/Workout/WorkoutEditor.swift
import Foundation
import SwiftData

/// Writes an edited draft back onto a finished workout, or deletes one, and keeps everything derived from it
/// consistent: set times on the records timeline, PR / set-record flags of every workout with those exercises,
/// attendance for the old and new day (locally and on the backend). Posts `.workoutHistoryDidChange` when done.
@MainActor
enum WorkoutEditor {

    static func apply(_ draft: WorkoutDraft, to workout: Workout, in context: ModelContext, today: Date = .now) {
        let cal = Calendar.current
        let oldStart = workout.startedAt
        let oldEnd = workout.endedAt ?? oldStart
        let oldDay = workout.completedSetCount > 0 ? cal.startOfDay(for: oldStart) : nil
        let exercisesBefore = exerciseIDs(of: workout)
        let times = WorkoutTimeline.completedTimes(for: draft, oldStart: oldStart, oldEnd: oldEnd)

        let name = draft.name.trimmingCharacters(in: .whitespacesAndNewlines)
        if !name.isEmpty { workout.name = name }
        workout.notes = draft.notes.trimmingCharacters(in: .whitespacesAndNewlines)
        workout.startedAt = draft.startedAt
        workout.endedAt = draft.endedAt

        // Exercises no longer in the draft; their sets go first (cascade is not reliable on iOS 17).
        let keptExercises = Set(draft.exercises.compactMap(\.sourceID))
        for entry in workout.exercises.filter({ !keptExercises.contains($0.persistentModelID) }) {
            for set in Array(entry.sets) { context.delete(set) }
            workout.exercises.removeAll { $0.persistentModelID == entry.persistentModelID }
            context.delete(entry)
        }

        for (index, item) in draft.exercises.enumerated() {
            guard let entry = entry(for: item, order: index, in: workout, context: context) else { continue }
            entry.order = index

            let keptSets = Set(item.sets.compactMap(\.sourceID))
            for set in entry.sets.filter({ !keptSets.contains($0.persistentModelID) }) {
                entry.sets.removeAll { $0.persistentModelID == set.persistentModelID }
                context.delete(set)
            }
            for (row, draftSet) in item.sets.enumerated() {
                let set: SetEntry
                if let id = draftSet.sourceID, let existing = entry.sets.first(where: { $0.persistentModelID == id }) {
                    set = existing
                } else {
                    set = SetEntry(order: row)
                    context.insert(set)
                    set.workoutExercise = entry
                }
                set.order = row
                set.kind = draftSet.kind
                set.weightKg = max(0, draftSet.weightKg)
                set.reps = max(0, draftSet.reps)
                set.completedAt = draftSet.isLogged ? (times[draftSet.id] ?? draft.startedAt) : nil
                if !draftSet.isLogged {
                    set.isPR = false
                    set.isSetRecord = false
                }
            }
        }
        try? context.save()

        RecordService.rebuild(exerciseIDs: exercisesBefore.union(exerciseIDs(of: workout)), in: context)
        let newDay = workout.completedSetCount > 0 ? cal.startOfDay(for: workout.startedAt) : nil
        let changes = AttendanceService.applyWorkoutDayChange(from: oldDay, to: newDay, excluding: workout.id,
                                                              context: context, today: today)
        try? context.save()
        reportAttendance(changes, in: context)
        NotificationCenter.default.post(name: .workoutHistoryDidChange, object: nil)
    }

    /// Deletes a finished workout. Call it once nothing shows the workout any more (after its sheet is gone).
    static func delete(_ workout: Workout, in context: ModelContext, today: Date = .now) {
        let ids = exerciseIDs(of: workout)
        let oldDay = workout.completedSetCount > 0 ? Calendar.current.startOfDay(for: workout.startedAt) : nil
        let workoutID = workout.id
        for entry in Array(workout.exercises) {
            for set in Array(entry.sets) { context.delete(set) }
            context.delete(entry)
        }
        context.delete(workout)
        try? context.save()

        RecordService.rebuild(exerciseIDs: ids, in: context)
        let changes = AttendanceService.applyWorkoutDayChange(from: oldDay, to: nil, excluding: workoutID,
                                                              context: context, today: today)
        try? context.save()
        reportAttendance(changes, in: context)
        NotificationCenter.default.post(name: .workoutHistoryDidChange, object: nil)
    }

    /// A workout in progress that no longer counts: "Edit sets" on the summary reopened it after its first Finish had
    /// marked the start day attended (and told the backend), every set was unticked, and it was then finished with
    /// nothing done or discarded. Reverts the day the way a delete does, unless another finished workout keeps it,
    /// and reports that. Nothing changes for a day that is not attended.
    static func releaseAttendance(of workout: Workout, in context: ModelContext, today: Date = .now) {
        let changes = AttendanceService.applyWorkoutDayChange(from: workout.startedAt, to: nil, excluding: workout.id,
                                                              context: context, today: today)
        guard !changes.isEmpty else { return }
        try? context.save()
        reportAttendance(changes, in: context)
    }

    // MARK: Helpers

    /// Tells the backend about the attendance the edit moved (attended on the new day, missed or planned again on
    /// the old one), the same way Finish reports a new workout.
    private static func reportAttendance(_ changes: [AttendanceService.WorkoutDayChange], in context: ModelContext) {
        guard !changes.isEmpty else { return }
        let schedule = AttendanceService.schedule(in: context)
        AttendanceSync.report(changes) { schedule?.isGymDay(Calendar.current.isoWeekday(for: $0)) ?? false }
    }

    private static func exerciseIDs(of workout: Workout) -> Set<String> {
        Set(workout.exercises.compactMap { $0.exercise?.id })
    }

    /// The row a draft exercise writes to: the one it came from, or a new one for an exercise added in the editor.
    private static func entry(for item: ExerciseDraft, order: Int, in workout: Workout,
                              context: ModelContext) -> WorkoutExercise? {
        if let id = item.sourceID, let existing = workout.exercises.first(where: { $0.persistentModelID == id }) {
            return existing
        }
        let exerciseID = item.exerciseID
        var d = FetchDescriptor<Exercise>(predicate: #Predicate { $0.id == exerciseID })
        d.fetchLimit = 1
        guard let exercise = try? context.fetch(d).first else { return nil }
        let entry = WorkoutExercise(order: order, exercise: exercise, restSeconds: item.restSeconds)
        context.insert(entry)
        entry.workout = workout
        return entry
    }
}

File: android/app/src/main/java/app/notomorrow/feature/workout/WorkoutEditor.kt
package app.notomorrow.feature.workout

import androidx.room.withTransaction
import app.notomorrow.data.dao.ExerciseDao
import app.notomorrow.data.dao.WorkoutDao
import app.notomorrow.data.entity.SetEntryEntity
import app.notomorrow.data.entity.WorkoutExerciseEntity
import app.notomorrow.data.relation.WorkoutWithExercises
import app.notomorrow.di.AppContainer
import app.notomorrow.service.AttendanceReporter
import app.notomorrow.service.AttendanceService
import app.notomorrow.service.RecordService
import app.notomorrow.util.Fmt
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import kotlin.math.max

/**
 * Writes an edited draft back onto a finished workout, or deletes one, and keeps everything derived
 * from it consistent — `WorkoutEditor.swift`: set times on the records timeline, PR / set-record
 * flags of every workout with those exercises, attendance for the old and new day.
 *
 * Everything runs in one transaction, so the Room flows (Train, Today, Progress, Bro) emit once,
 * with the whole edit applied — the Android stand-in for `.workoutHistoryDidChange`.
 */
object WorkoutEditor {

    /**
     * What a save or delete touches. [transaction] is `db.withTransaction` in the app, a plain call
     * in tests; [reportAttendance] sends the attendance it changed to the backend once the
     * transaction has landed (`AttendanceSync.report(changes)`).
     */
    class Stores(
        val workoutDao: WorkoutDao,
        val exerciseDao: ExerciseDao,
        val recordService: RecordService,
        val attendanceService: AttendanceService,
        val transaction: suspend (suspend () -> Unit) -> Unit = { it() },
        val reportAttendance: AttendanceReporter = AttendanceReporter.None,
    ) {
        companion object {
            fun of(container: AppContainer) = Stores(
                workoutDao = container.db.workoutDao(),
                exerciseDao = container.db.exerciseDao(),
                recordService = container.recordService,
                attendanceService = container.attendanceService,
                transaction = { block -> container.db.withTransaction { block() } },
                reportAttendance = container.attendanceReporter,
            )
        }
    }

    /**
     * `apply(_:to:in:today:)`, plan §5.2: name (an empty one keeps the old), notes, start and end;
     * exercises and sets removed, re-indexed, updated or inserted; each logged row's time remapped
     * or borrowed from a neighbour, open rows cleared with their flags; then the records rebuilt
     * for the exercises before ∪ after, and attendance corrected for the day it counted on before
     * and after. Returns `false` when the workout is gone.
     */
    suspend fun save(
        workoutId: String,
        draft: WorkoutDraft,
        stores: Stores,
        zone: ZoneId = ZoneId.systemDefault(),
        today: LocalDate = LocalDate.now(zone),
    ): Boolean {
        var saved = false
        var changes = emptyList<AttendanceService.WorkoutDayChange>()
        stores.transaction body@{
            val dao = stores.workoutDao
            val graph = dao.workoutWithExercises(workoutId) ?: return@body
            val workout = graph.workout
            val oldStart = workout.startedAt
            val oldEnd = workout.endedAt ?: oldStart
            val oldDay = countedDay(graph, zone)
            val exercisesBefore = exerciseIds(graph)
            val times = WorkoutTimeline.completedTimes(draft, oldStart, oldEnd)

            val name = draft.name.trim()
            dao.updateWorkout(
                workout.copy(
                    name = name.ifEmpty { workout.name },
                    notes = draft.notes.trim(),
                    startedAt = draft.startedAt,
                    endedAt = draft.endedAt,
                ),
            )

            // Exercises no longer in the draft; their sets go with them (foreign-key cascade).
            val keptExercises = draft.exercises.mapNotNull { it.sourceId }.toSet()
            graph.exercises
                .filter { it.workoutExercise.id !in keptExercises }
                .forEach { dao.deleteWorkoutExercise(it.workoutExercise) }

            draft.exercises.forEachIndexed { index, item ->
                val existing = graph.exercises.firstOrNull { it.workoutExercise.id == item.sourceId }
                val entryId = when {
                    existing != null -> {
                        if (existing.workoutExercise.order != index) {
                            dao.updateWorkoutExercise(existing.workoutExercise.copy(order = index))
                        }
                        existing.workoutExercise.id
                    }
                    // An exercise gone from the library since it was picked is skipped, as iOS does.
                    stores.exerciseDao.byId(item.exerciseId) == null -> return@forEachIndexed
                    else -> dao.insertWorkoutExercise(
                        WorkoutExerciseEntity(
                            workoutId = workoutId,
                            exerciseId = item.exerciseId,
                            order = index,
                            restSeconds = item.restSeconds,
                        ),
                    )
                }

                val keptSets = item.sets.mapNotNull { it.sourceId }.toSet()
                existing?.sets?.filter { it.id !in keptSets }?.forEach { dao.deleteSet(it) }
                item.sets.forEachIndexed { row, set ->
                    val completedAt = if (set.isLogged) times[set.id] ?: draft.startedAt else null
                    val source = existing?.sets?.firstOrNull { it.id == set.sourceId }
                    if (source != null) {
                        val updated = source.copy(
                            order = row,
                            kind = set.kind,
                            weightKg = max(0.0, set.weightKg),
                            reps = max(0, set.reps),
                            completedAt = completedAt,
                            isPR = if (set.isLogged) source.isPR else false,
                            isSetRecord = if (set.isLogged) source.isSetRecord else false,
                        )
                        if (updated != source) dao.updateSet(updated)
                    } else {
                        dao.insertSet(
                            SetEntryEntity(
                                workoutExerciseId = entryId,
                                order = row,
                                kind = set.kind,
                                weightKg = max(0.0, set.weightKg),
                                reps = max(0, set.reps),
                                completedAt = completedAt,
                            ),
                        )
                    }
                }
            }

            val after = dao.workoutWithExercises(workoutId) ?: return@body
            stores.recordService.rebuild(exercisesBefore + exerciseIds(after))
            changes = correctAttendance(oldDay, countedDay(after, zone), workoutId, stores, zone, today)
            saved = true
        }
        report(changes, stores)
        return saved
    }

    /**
     * `delete(_:in:today:)` — deletes a finished workout (its exercises and sets cascade), rebuilds
     * the records of its exercises and corrects the day it counted on. Call it once nothing shows the
     * workout any more (its sheet is closing).
     */
    suspend fun delete(
        workoutId: String,
        stores: Stores,
        zone: ZoneId = ZoneId.systemDefault(),
        today: LocalDate = LocalDate.now(zone),
    ) {
        var changes = emptyList<AttendanceService.WorkoutDayChange>()
        stores.transaction body@{
            val graph = stores.workoutDao.workoutWithExercises(workoutId) ?: return@body
            val ids = exerciseIds(graph)
            val oldDay = countedDay(graph, zone)
            stores.workoutDao.deleteWorkoutById(workoutId)
            stores.recordService.rebuild(ids)
            changes = correctAttendance(oldDay, null, workoutId, stores, zone, today)
        }
        report(changes, stores)
    }

    // MARK: - Helpers

    private fun exerciseIds(graph: WorkoutWithExercises): Set<String> =
        graph.exercises.mapNotNull { it.workoutExercise.exerciseId }.toSet()

    /** The day a workout counts on for attendance: its start day, when it has a completed set. */
    internal fun countedDay(graph: WorkoutWithExercises, zone: ZoneId): LocalDate? =
        if (graph.completedSetCount > 0) Instant.ofEpochMilli(graph.workout.startedAt).atZone(zone).toLocalDate() else null

    private suspend fun correctAttendance(
        oldDay: LocalDate?,
        newDay: LocalDate?,
        workoutId: String,
        stores: Stores,
        zone: ZoneId,
        today: LocalDate,
    ): List<AttendanceService.WorkoutDayChange> =
        correctAttendance(oldDay, newDay, workoutId, stores.workoutDao, stores.attendanceService, zone, today)

    /**
     * `AttendanceService.applyWorkoutDayChange(from:to:excluding:)`: moves the day workout
     * [workoutId] counts on from [oldDay] to [newDay] (`null` = it no longer counts), keeping
     * [oldDay] attended while another finished workout with a completed set started on it. Also
     * the active workout's Finish with nothing done, and its Discard.
     */
    internal suspend fun correctAttendance(
        oldDay: LocalDate?,
        newDay: LocalDate?,
        workoutId: String,
        workoutDao: WorkoutDao,
        attendanceService: AttendanceService,
        zone: ZoneId,
        today: LocalDate,
    ): List<AttendanceService.WorkoutDayChange> {
        if (oldDay == newDay) return emptyList()
        val stillAttended = oldDay?.let { day ->
            val from = day.atStartOfDay(zone).toInstant().toEpochMilli()
            val to = day.plusDays(1).atStartOfDay(zone).toInstant().toEpochMilli()
            workoutDao.countedWorkoutsBetween(from, to, excludingId = workoutId) > 0
        } ?: false
        return attendanceService.applyWorkoutDayChange(oldDay, newDay, stillAttended, today)
    }

    private suspend fun report(changes: List<AttendanceService.WorkoutDayChange>, stores: Stores) =
        report(changes, stores.attendanceService, stores.reportAttendance)

    /**
     * Sends what [correctAttendance] wrote to the backend (`AttendanceService.wireStatus`): the new
     * day attended, a past gym day missed, a gym day given back planned. A cleared rest day has no
     * server equivalent.
     */
    internal suspend fun report(
        changes: List<AttendanceService.WorkoutDayChange>,
        attendanceService: AttendanceService,
        reporter: AttendanceReporter,
    ) {
        if (changes.isEmpty()) return
        val schedule = attendanceService.schedule()
        reporter.report(changes) { schedule?.isGymDay(Fmt.isoWeekday(it)) ?: false }
    }
}
